### DISCUSSION TITLE
Progress bar & spinning wheel widgets

### DISCUSSION THREAD

**Discussion by mehrdadfeller** [ubopod/ubo_app#74] [Ideas]
In certain interactions such as fetching a docker image, and generally waiting for a process to finish, we need to define progress bar & spinning wheel widgets. The UI logic for progress bar is that it takes a float between 0 and 1 and sets the bar. The spinning wheel only spins freely with no "up to" point.

We need to have some business logics baked into these widget that offers a high-level implementation for specific applications such as download progress bar. I guess for download we should show some estimated download speed if there's an easy way for it. 

Currently the user may feel a bit lost and not so sure if the download is indeed in progress or it is just stuck.

**Comment by zytegalaxy** [comment]
@sassanh perhaps we can break this into smaller items / issues?

**Reply by sassanh** [reply]
For the docker image pull action, after considering the challenges, our conclusion was:
Create an issue for implementing the most sophisticated solution which is consuming the json stream of the docker client, getting the metadata of the image and the description of its layers from the hub (or maybe it is already present locally when the image starts downloading), getting the size of each layer from the hub, get an initial status of each layer as the stream may never report any status for already downloaded layers and eventually do a weighted average on the progressions of different layers to extract a final single progression value. We may implement it in a completely different repository.

On the other hand, as a general solution for pending states, we can consider:
1. A progress bar widget in ubo-gui
2. A spinning widget in ubo-gui
3. An LED animation for the waiting status.
4. An LED animation for the progress bar status.

@zytegalaxy Is there anything missing? If not I will create the issues.

**Reply by sassanh** [reply]
Related: https://github.com/ubopod/ubo-gui/issues/26

**Reply by sassanh** [reply]
Sample docker json stream:
[docker-pull-stream.json](https://github.com/ubopod/ubo_app/files/15144008/docker-pull-stream.json)
This is generated with something like this:
```python
import docker

client = docker.from_env()
output = open('docker-pull-stream.json', 'w')
response = client.api.pull('nginx/nginx-ingress:latest',stream=True,decode=True)
for line in response:
  output.write(json.dumps(line) + '\n')
output.close()
```

**Reply by sassanh** [reply]
We also need a general "download" helper which takes a url of a resource and a destination, and utilizes the progress bar widget to demonstrate the progress of download.
We can use pycurl and its [progress callback](http://pycurl.io/docs/latest/callbacks.html) for this purpose.

**Comment by mehrdadfeller** [comment]
Other places that can benefit from a progress bar:

- Ubo app update process (pulling a new update and installing it)
- VSCode CLI download
- Any other non-docker file download process

**Comment by sassanh** [comment]
We need to have an abstraction for async tasks with progression, some suggestions:
1. Using the generator functions of Python, a bit tricky to keep track of the async job and cancel it and take care of memory leaks, etc.
2. Using python-redux, needs a bit of benchmarking to see if the load is beyond the capacity or not. If python-redux can handle such load it is the best option as the progress information will be available in the main bus and everyone can consume it.
3. A subscribable value like autorun's return value, but not in the python-redux bus.

I will do some benchmarking on python-redux and will report back here.
